File: litho-it/src/test/java/com/facebook/litho/LayoutStateSizeTest.java
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho;

import static androidx.test.core.app.ApplicationProvider.getApplicationContext;
import static org.assertj.core.api.Java6Assertions.assertThat;

import com.facebook.litho.testing.TestLayoutComponent;
import com.facebook.litho.testing.Whitebox;
import com.facebook.litho.testing.testrunner.ComponentsTestRunner;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;

@RunWith(ComponentsTestRunner.class)
public class LayoutStateSizeTest {
  private static final int COMPONENT_ID = 37;
  private static final int WIDTH = 49;
  private static final int HEIGHT = 51;

  private LayoutState mLayoutState;
  private Component mComponent;
  private ComponentContext mContext;

  @Before
  public void setup() {
    mContext = new ComponentContext(getApplicationContext());
    mComponent = TestLayoutComponent.create(mContext).build();
    Whitebox.setInternalState(mComponent, "mId", COMPONENT_ID);

    mLayoutState = new LayoutState(mContext);
    Whitebox.setInternalState(mLayoutState, "mWidth", WIDTH);
    Whitebox.setInternalState(mLayoutState, "mHeight", HEIGHT);
    Whitebox.setInternalState(mLayoutState, "mComponent", mComponent);
  }

  @Test
  public void testCompatibleSize() {
    assertThat(mLayoutState.isCompatibleSize(WIDTH, HEIGHT)).isTrue();
  }

  @Test
  public void testIncompatibleWidthSpec() {
    assertThat(mLayoutState.isCompatibleSize(WIDTH + 1000, HEIGHT)).isFalse();
  }

  @Test
  public void testIncompatibleHeightSpec() {
    assertThat(mLayoutState.isCompatibleSize(WIDTH, HEIGHT + 1000)).isFalse();
  }
}

File: litho-it/src/test/java/com/facebook/litho/LayoutStateSizeTest.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.litho

import android.content.Context
import androidx.test.core.app.ApplicationProvider
import com.facebook.litho.SizeSpec.makeSizeSpec
import com.facebook.litho.testing.TestLayoutComponent
import com.facebook.litho.testing.Whitebox
import com.facebook.litho.testing.testrunner.LithoTestRunner
import org.assertj.core.api.Assertions.assertThat
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(LithoTestRunner::class)
class LayoutStateSizeTest {

  private lateinit var layoutState: LayoutState
  private lateinit var component: Component
  private lateinit var context: ComponentContext

  @Before
  fun setup() {
    context = ComponentContext(ApplicationProvider.getApplicationContext<Context>())
    component = TestLayoutComponent.create(context).build()
    Whitebox.setInternalState(component, "mId", COMPONENT_ID)
    val result =
        ResolveTreeFuture.resolve(context, component, TreeState(), 1, 1, null, null, null, null)
    layoutState =
        LayoutTreeFuture.layout(
            result,
            makeSizeSpec(WIDTH, SizeSpec.EXACTLY),
            makeSizeSpec(HEIGHT, SizeSpec.EXACTLY),
            1,
            1,
            false,
            null,
            null,
            null,
            null)
  }

  @Test
  fun testCompatibleSize() {
    assertThat(layoutState.isCompatibleSize(WIDTH, HEIGHT)).isTrue
  }

  @Test
  fun testIncompatibleWidthSpec() {
    assertThat(layoutState.isCompatibleSize(WIDTH + 1_000, HEIGHT)).isFalse
  }

  @Test
  fun testIncompatibleHeightSpec() {
    assertThat(layoutState.isCompatibleSize(WIDTH, HEIGHT + 1_000)).isFalse
  }

  companion object {
    private const val COMPONENT_ID = 37
    private const val WIDTH = 49
    private const val HEIGHT = 51
  }
}
